db_test.go

```diff
@@ -1303,6 +1303,100 @@ func TestInitializeHeadTimestamp(t *testing.T) {
 	})
 }
 
+func TestNoEmptyBlocks(t *testing.T) {
+	db, close := openTestDB(t, nil)
+	defer close()
+	defer db.Close()
+	db.DisableCompactions()
+
+	// Test no blocks after compact with empty head.
+	testutil.Ok(t, db.compact())
+	testutil.Equals(t, 0, len(db.Blocks()))
+
+	// Test no blocks after deleting all samples from head.
+	blockRange := DefaultOptions.BlockRanges[0]
+	label := labels.FromStrings("foo", "bar")
+
+	app := db.Appender()
+	for i := int64(0); i < 6; i++ {
+		_, err := app.Add(label, i*blockRange, 0)
+		testutil.Ok(t, err)
+		_, err = app.Add(label, i*blockRange+1000, 0)
+		testutil.Ok(t, err)
+	}
+	testutil.Ok(t, app.Commit())
+
+	testutil.Ok(t, db.Delete(math.MinInt64, math.MaxInt64, labels.NewEqualMatcher("foo", "bar")))
+	uid, err := db.compactor.Write(db.dir, db.head, db.head.MinTime(), db.head.MaxTime(), nil)
+	testutil.Ok(t, err)
+	testutil.Equals(t, ulid.ULID{}, uid)
+	testutil.Ok(t, db.reload())
+	testutil.Ok(t, db.head.Truncate(db.head.MaxTime()))
+	// No blocks created.
+	testutil.Equals(t, 0, len(db.Blocks()))
```

Comment:
don't you think it would be better to actually check the blocks on disk rather than relying on db.Blocks()

```
bb, err := blockDirs(db.Dir())
testutil.Ok(t, err)
testutil.Equals(t, 0, len(bb))
```

Reply:
I think `db.reload()` takes care of that. `db.blocks` is updated with the blocks on disk in `db.reload()`.
https://github.com/prometheus/tsdb/blob/38a2c6b6bd454443896a401c236fdea8383441d5/db.go#L534-L554

Reply:
if there is a bug in db.reload and Blocks() might return incorrect blocks as well.